base/logger_audit.go

```diff
@@ -86,20 +86,45 @@ func expandFields(id AuditID, ctx context.Context, additionalData AuditFields) A
 
 	fields[auditFieldTimestamp] = time.Now()
 
-	// TODO: CBG-3976 - Inject and merge data from env var
+	fields.merge(ctx, globalFields, isDevMode)
 	// TODO: CBG-3977 - Inject and merge data from request header
 
 	return fields
 }
 
+// Merge will perform a shallow overwrite of the fields in the AuditFields. If there are conflicts, do not overwrite but log a warning. This will panic in dev mode.
+func (f *AuditFields) merge(ctx context.Context, overwrites AuditFields, isDevMode bool) {
+	var duplicateFields []string
+	for k, v := range overwrites {
+		_, ok := (*f)[k]
+		if ok {
+			duplicateFields = append(duplicateFields, fmt.Sprintf("%q=%q", k, v))
+			continue
+		}
+		(*f)[k] = v
+	}
+	if duplicateFields != nil {
+		msg := fmt.Sprintf("audit fields %s already exist in base audit fields %+v, will not overwrite an audit event", strings.Join(duplicateFields, ","), *f)
+		if isDevMode {
+			panic(msg)
+		}
+		WarnfCtx(ctx, msg)
+	}
+}
+
 // Audit creates and logs an audit event for the given ID and a set of additional data associated with the request.
 func Audit(ctx context.Context, id AuditID, additionalData AuditFields) {
+	audit(ctx, id, additionalData, IsDevMode())
```

Comment:
Do these new layers of indirection for `isDevMode` retain the benefit of the compiler removing devmode branches when not enabled (see #6898)? That was one of my largest concerns when adding it - make evaluation of it completely free for non-dev mode.

As far as I can tell the only benefit to parameterising `isDevMode` is to get test coverage of the warning? Is that really necessary? It clutters up the API quite a bit and introduces the above concern.

Reply:
I think I should just drop panics because they are unlikely to happen in tests and there's more benefit to actually testing the production behavior than a panic to stop the behavior that isn't really going to be seen without capella's usage of it. This might be different for request handling vs global handling though.